### PR TITLE
many: check for apparmor prompting support when enabling the experimental feature

### PR DESCRIPTION
**Pull request by bboozzoo**
When attempting to enable apparmor prompting experimental feature, first check whether it is supported by the system at all.

Related: SNAPDENG-22972

~~Note that this basically enables the prompting feature if both kernel and parser support it. Perhaps we should still gate this with an environment flag, eg SNAPD_APPARMOR_PROMPTING_ENABLE=1 ?~~

### CODE REVIEW

File: features/features.go [hunk 1]
```diff
@@ -221,14 +221,20 @@ func (f SnapdFeature) ConfigOption() (snapName, confName string) {
...
+func (f SnapdFeature) IsSupported() (supported bool, whyNot string) {
 	if callback, exists := featuresSupportedCallbacks[f]; exists {
-		supported, _ := callback() // discard unsupported reason
-		return supported
+		supported, whyNot = callback() // discard unsupported reason
+		if supported {
+			// discard the reason if the feature is already
+			// supported
+			return true, ""
+		}
```
Comment (ZeyadYasser) [inline features/features.go:234]:
```suggestion
		if !supported {
			return false, whyNot
		}
```

File: overlord/configstate/configcore/prompting.go [hunk 2]
```diff
@@ -49,6 +51,16 @@ func doExperimentalApparmorPromptingDaemonRestart(c RunTransaction, opts *fsOnly
 		return nil
 	}
 
+	if prompting {
+		if is, whyNot := features.AppArmorPrompting.IsSupported(); !is {
+			if whyNot == "" {
+				// we have details as to why
```
Comment (ZeyadYasser) [inline overlord/configstate/configcore/prompting.go:57]:
```suggestion
				// we don't have details as to why
```

**Review by ZeyadYasser — APPROVED** [review]
Thank you!

File: overlord/configstate/configcore/prompting.go (hunk 2 above)
Comment (bboozzoo) [reply overlord/configstate/configcore/prompting.go:57]: haha, good catch, thank you!

File: overlord/configstate/configcore/prompting.go [hunk 3]
```diff
@@ -49,6 +51,16 @@ func doExperimentalApparmorPromptingDaemonRestart(c RunTransaction, opts *fsOnly
 		return nil
 	}
```
Comment (pedronis) [inline overlord/configstate/configcore/prompting.go:53]: for now at least we should probably refuse to turn this on on Ubuntu Core

File: overlord/configstate/configcore/prompting.go [hunk 4]
```diff
@@ -49,6 +51,16 @@ func doExperimentalApparmorPromptingDaemonRestart(c RunTransaction, opts *fsOnly
...
 
+	if prompting {
+		if is, whyNot := features.AppArmorPrompting.IsSupported(); !is {
+			if whyNot == "" {
+				// we don't have details as to why
+				return fmt.Errorf("prompting feature is not supported by the system")
```
Comment (pedronis) [inline overlord/configstate/configcore/prompting.go:58]: maybe "cannot enable prompting feature as it is not ...."

File: overlord/configstate/configcore/prompting.go [hunk 5]
```diff
@@ -49,6 +51,16 @@ func doExperimentalApparmorPromptingDaemonRestart(c RunTransaction, opts *fsOnly
...
+			}
+			return fmt.Errorf("prompting feature is not supported by the system, reason: %s", whyNot)
```
Comment (pedronis) [inline overlord/configstate/configcore/prompting.go:60]: same

**Review by pedronis — COMMENTED** [review]
thank you, some suggestions

File: overlord/configstate/configcore/prompting.go [hunk 6]
```diff
@@ -52,12 +53,16 @@ func doExperimentalApparmorPromptingDaemonRestart(c RunTransaction, opts *fsOnly
 	}
 
 	if prompting {
+		if !release.OnClassic && !release.OnCoreDesktop {
```
Comment (pedronis) [inline overlord/configstate/configcore/prompting.go:56]: please leave a TODO about doing the right thing for preseeding

**Review by pedronis — APPROVED** [review]
thanks, one small comment

File: sandbox/apparmor/apparmor_test.go [hunk 7]
```diff
@@ -645,11 +645,8 @@ func (s *apparmorSuite) TestPromptingSupported(c *C) {
 	defer restore()
 
 	supported, reason := apparmor.PromptingSupported()
-	// TODO: change this once snapd supports prompting
-	c.Check(supported, Equals, false)
-	c.Check(reason, Equals, "requires newer version of snapd")
-	// c.Check(supported, Equals, true)
-	// c.Check(reason, Equals, "")
+	c.Check(supported, Equals, true)
+	c.Check(reason, Equals, "")
```
Comment (olivercalder) [inline sandbox/apparmor/apparmor_test.go:649]: Doing this here means the capstone PR can be closed: https://github.com/canonical/snapd/pull/14318/commits/b9653eaff9e35c4fd14e094ae64ec492eebe4eb8

I agree it's better to do this now, that way we can test more thoroughly now and later in the PR chain. Only worry is if someone running edge manually enables prompting, but that's unlikely.

**Review by olivercalder — COMMENTED** [review]
One question about the location of checks for being on classic or core desktop. Otherwise, looks good!

File: overlord/configstate/configcore/prompting.go [hunk 8]
```diff
@@ -49,6 +52,20 @@ func doExperimentalApparmorPromptingDaemonRestart(c RunTransaction, opts *fsOnly
 		return nil
 	}
 
+	if prompting {
+		if !release.OnClassic && !release.OnCoreDesktop {
+			return fmt.Errorf("cannot enable prompting feature as it is not supported on Ubuntu Core systems")
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply overlord/configstate/configcore/prompting.go:57]: Makes sense. What about in `features.AppArmorPrompting.IsSupported()?

[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/configstate/configcore/prompting.go:59]: hm maybe, why not.  I suppose we can do it in a followup after we get this one landed.

**Review by olivercalder — APPROVED** [review]
Looks good to me, thanks! I'm ambivalent whether the not-core checks should be moved to `features.AppArmorPrompting.IsSupported` (in particular, the [callback](https://github.com/canonical/snapd/pull/14340/files#diff-a8b6135d50534471326ea7bcd20e0f5eae25353f7788338060f718128a6a0b34R176)) now, or in a follow-up. Change should be simple enough either way.

File: overlord/configstate/configcore/prompting.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/configstate/configcore/prompting.go:59]: the check should probably stay here until we deal with the TODO anyway, that is a bit of a more complex question

File: features/features.go [hunk 9]
```diff
@@ -221,14 +221,17 @@ func (f SnapdFeature) ConfigOption() (snapName, confName string) {
...
-// or if it has no supportedCallback.
-func (f SnapdFeature) IsSupported() bool {
+// IsSupported returns true if the feature's supported callback returns true, or
+// if it has no supportedCallback. If the feature is unsupported, the returned
+// string details as to why.
+func (f SnapdFeature) IsSupported() (supported bool, whyNot string) {
 	if callback, exists := featuresSupportedCallbacks[f]; exists {
-		supported, _ := callback() // discard unsupported reason
-		return supported
+		supported, whyNot = callback() // discard unsupported reason
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply features/features.go:229]: The `// discard unsupported reason` should definitely be removed

**Comment by bboozzoo** [comment]
This can be force merged now.

**Comment by bboozzoo** [comment]
Added skip spread label, will close and reopen to get this landed.

**Pull request merged**